org.jacoco.core/src/org/jacoco/core/internal/analysis/MethodCoverageCalculator.java

```diff
@@ -0,0 +1,159 @@
+ * Copyright (c) 2009, 2018 Mountainminds GmbH & Co. KG and Contributors
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Evgeny Mandrikov - initial API and implementation
+ *    
+ *******************************************************************************/
+package org.jacoco.core.internal.analysis;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.jacoco.core.analysis.ISourceFileCoverage;
+import org.jacoco.core.analysis.ISourceNode;
+import org.jacoco.core.internal.analysis.filter.IFilterOutput;
+import org.objectweb.asm.tree.AbstractInsnNode;
+
+/**
+ * Calculates the filtered coverage of a single method. A instance of this class
+ * can be first used as {@link IFilterOutput} before the coverage result is
+ * calculated.
+ */
+class MethodCoverageCalculator implements IFilterOutput {
+
+	private final Map<AbstractInsnNode, Instruction> instructions;
+
+	private final Set<AbstractInsnNode> ignored;
+
+	/**
+	 * Instructions that should be merged form disjoint sets. Coverage
+	 * information from instructions of one set will be merged into
+	 * representative instruction of set.
+	 * 
+	 * Each such set is represented as a singly linked list: each element except
+	 * one references another element from the same set, element without
+	 * reference - is a representative of this set.
+	 * 
+	 * This map stores reference (value) for elements of sets (key).
+	 */
+	private final Map<AbstractInsnNode, AbstractInsnNode> merged;
+
+	private final Map<AbstractInsnNode, Set<AbstractInsnNode>> replacements;
+
+	MethodCoverageCalculator(
+			final Map<AbstractInsnNode, Instruction> instructions) {
+		this.instructions = instructions;
+		this.ignored = new HashSet<AbstractInsnNode>();
+		this.merged = new HashMap<AbstractInsnNode, AbstractInsnNode>();
+		this.replacements = new HashMap<AbstractInsnNode, Set<AbstractInsnNode>>();
+	}
+
+	/**
+	 * Applies all specified filtering commands and calculates the resulting
+	 * coverage.
+	 * 
+	 * @param coverage
+	 *            the result is added to this coverage node
+	 */
+	void calculate(final MethodCoverageImpl coverage) {
+
+		// Merge and calculate line range:
+		int firstLine = ISourceFileCoverage.UNKNOWN_LINE;
+		int lastLine = ISourceFileCoverage.UNKNOWN_LINE;
+
+		for (final Map.Entry<AbstractInsnNode, Instruction> entry : instructions
+				.entrySet()) {
+
+			final Instruction i = entry.getValue();
+			final AbstractInsnNode n = entry.getKey();
+			final AbstractInsnNode r = findRepresentative(n);
+			if (r != n) {
+				ignored.add(n);
+				instructions.put(r, instructions.get(r).merge(i));
```

Comment:
To me seems that something is wrong here.

Before this change we were performing merges in two steps - see #747
1. from all instructions into their representative instruction
2. from all representative instructions back

and seems that this is gone.

Consider following case for `MethodCoverageCalculatorTest`, whose assertion currently will fail

```java
	@Test
	public void test() {
		InsnNode i1 = addInsn(1, false);
		InsnNode i2 = addInsn(2, false);
		InsnNode i3 = addInsn(2, true);

		MethodCoverageCalculator c = new MethodCoverageCalculator(instructions);
		c.merge(i3, i2);
		c.replaceBranches(i1, Collections.<AbstractInsnNode> singleton(i2));
		c.calculate(coverage);

		assertLine(1, 0, 1, 0, 0);
	}
```

where

1. initially `i1` and `i2` was not covered, `i3` was covered
2. merge `i3` and `i2` - both should be considered as covered since `i3` was covered
3. replace `i1` by `i2` - `i1` should be considered as covered because of previous merge

This test fails even if `c.merge(i3, i2)` replaced by

```
c.merge(i3, i2);
c.merge(i2, i3);
```

and passes when replaced by

```
c.merge(i2, i3);
```

or by

```
c.merge(i2, i3);
c.merge(i3, i2);
```

Reply:
@Godin Excellent catch! I added your test case and applied a slightly different fix: Instead of iterating over all instructions every time I think we can pick the representative instruction when replacing branches.

Reply:
@marchof 

> pick the representative instruction when replacing branches

IMO this might become source of errors in future. What do you think about replacement of

```
instructions.put(r, instructions.get(r).merge(i));
```

by

```
instruction = instructions.get(representativeNode).merge(instruction);
instructions.put(r, instruction);
instructions.put(n, instruction);
```

instead?

Reply:
Looks good to me at the first glance and all tests are green (if I remove the lookup during handling replacements).

But: Not sure whether this is correct in case we merge more than two instructions. If we merge pairwise with the representative will the total merge result be propagated to all instructions of the group? Trying to create a test case now...

Reply:
@marchof representative is the same

Reply:
@marchof this will work if we mutate instructions, but now they are immutable so probably you right and this won't work :(

Reply:
Here is a demonstration of the issue:

	@Test
	public void should_replace_branches_with_merged_instructions() {
		InsnNode i1 = addInsn(1, false, false, false);
		InsnNode i2 = addInsn(2, true);
		InsnNode i3 = addInsn(2, false);
		InsnNode i4 = addInsn(2, false);

		MethodCoverageCalculator c = new MethodCoverageCalculator(instructions);
		c.merge(i4, i3);
		c.merge(i3, i2);
		c.replaceBranches(i1,
				new HashSet<AbstractInsnNode>(Arrays.asList(i2, i3, i4)));
		c.calculate(coverage);

		assertLine(1, 0, 1, 0, 3);
	}

Works with current version, but breaks without representative lookup during replacement.